qiskit_dynamics/pulse/string_model_parser/string_model_parser.py

```diff
@@ -0,0 +1,305 @@
+            else:
+                static_hamiltonian += op
+        else:
+            channel = channel.lower()
+            if channel in reduced_channels:
+                hamiltonian_operators[reduced_channels.index(channel)] += op
+            else:
+                hamiltonian_operators.append(op)
+                reduced_channels.append(channel)
+
+    # sort channels/operators according to channel ordering
+    if len(reduced_channels) > 0:
+        reduced_channels, hamiltonian_operators = zip(
+            *sorted(zip(reduced_channels, hamiltonian_operators))
+        )
+
+    return static_hamiltonian, list(hamiltonian_operators), list(reduced_channels), subsystem_dims
+
+
+def hamiltonian_pre_parse_exceptions(hamiltonian_dict: dict):
+    """Raises exceptions for improperly formatted or unsupported elements of
+    hamiltonian dict specification.
+
+    Parameters:
+        hamiltonian_dict: Dictionary specification of hamiltonian.
+    Returns:
+    Raises:
+        QiskitError: If some part of the Hamiltonian dictionary is unsupported or invalid.
+    """
+
+    ham_str = hamiltonian_dict.get("h_str", [])
+    if ham_str in ([], [""]):
+        raise QiskitError("Hamiltonian dict requires a non-empty 'h_str' entry.")
+
+    if hamiltonian_dict.get("qub", {}) == {}:
+        raise QiskitError(
+            "Hamiltonian dict requires non-empty 'qub' entry with subsystem dimensions."
+        )
+
+    if hamiltonian_dict.get("osc", {}) != {}:
+        raise QiskitError("Oscillator-type systems are not supported.")
+
+    # verify that if terms in h_str have the divider ||, then the channels are in the valid format
+    for term in hamiltonian_dict["h_str"]:
+        malformed_text = """Term '{}' does not conform to required string format.
+                            Channels may only be specified in the format
+                            'aa||Cxx', where 'aa' specifies an operator,
+                            C is a valid channel character,
+                            and 'xx' is a string of digits.""".format(
+            term
+        )
+
+        # if two vertical bars used together, check if channels in correct format
+        if term.count("|") == 2 and term.count("||") == 1:
+            # get the string reserved for channel
+            channel_str = term[term.index("||") + 2 :]
+
+            # if channel string is empty
+            if len(channel_str) == 0:
+                raise QiskitError(malformed_text)
+
+            # if first entry in channel string isn't a valid channel character
+            if channel_str[0] not in CHANNEL_CHARS:
+                raise QiskitError(malformed_text)
+
+            # Verify either that: all remaining characters are digits, or,
+            # if term starts with _SUM[ and ends with ], all remaining characters
+            # are either digits, or starts and ends with {}
+            if term[-1] == "]" and len(term) > 5 and term[:5] == "_SUM[":
```

Comment:
Shouldn't we do checks like these even on strings without channels? (specifically line 280)

Reply:
It appears all of these checks are based around validating channel specification - nothing else is being checked here. I think I just did this here because the legacy parser has some of its own checking for the validity of the operator part of the string, but has less checking for the validity of the channel part.